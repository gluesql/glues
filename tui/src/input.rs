use tui_textarea::Input as TextAreaInput;

#[derive(Clone, Debug)]
pub enum Input {
    Key(KeyEvent),
    Paste(String),
    Resize(u16, u16),
}

#[derive(Clone, Debug)]
pub struct KeyEvent {
    pub code: KeyCode,
    pub modifiers: KeyModifiers,
    pub kind: KeyEventKind,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum KeyCode {
    Char(char),
    F(u8),
    Backspace,
    Enter,
    Left,
    Right,
    Up,
    Down,
    Tab,
    Delete,
    Home,
    End,
    PageUp,
    PageDown,
    Esc,
    Null,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum KeyEventKind {
    Press,
    Release,
    Repeat,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct KeyModifiers {
    pub ctrl: bool,
    pub alt: bool,
    pub shift: bool,
}

impl KeyModifiers {
    pub const NONE: Self = Self {
        ctrl: false,
        alt: false,
        shift: false,
    };

    pub const CONTROL: Self = Self {
        ctrl: true,
        alt: false,
        shift: false,
    };

    pub fn new(ctrl: bool, alt: bool, shift: bool) -> Self {
        Self { ctrl, alt, shift }
    }
}

impl Default for KeyModifiers {
    fn default() -> Self {
        Self::NONE
    }
}

#[cfg(not(target_arch = "wasm32"))]
mod native {
    use super::*;
    use ratatui::crossterm::event as ct;

    impl From<ct::Event> for Input {
        fn from(event: ct::Event) -> Self {
            match event {
                ct::Event::Key(key) => Input::Key(key.into()),
                ct::Event::Paste(paste) => Input::Paste(paste),
                ct::Event::Resize(w, h) => Input::Resize(w, h),
                _ => Input::Resize(0, 0),
            }
        }
    }

    impl From<ct::KeyEvent> for KeyEvent {
        fn from(key: ct::KeyEvent) -> Self {
            let modifiers = KeyModifiers::new(
                key.modifiers.contains(ct::KeyModifiers::CONTROL),
                key.modifiers.contains(ct::KeyModifiers::ALT),
                key.modifiers.contains(ct::KeyModifiers::SHIFT),
            );

            let code = match key.code {
                ct::KeyCode::Char(c) => KeyCode::Char(c),
                ct::KeyCode::F(n) => KeyCode::F(n),
                ct::KeyCode::Backspace => KeyCode::Backspace,
                ct::KeyCode::Enter => KeyCode::Enter,
                ct::KeyCode::Left => KeyCode::Left,
                ct::KeyCode::Right => KeyCode::Right,
                ct::KeyCode::Up => KeyCode::Up,
                ct::KeyCode::Down => KeyCode::Down,
                ct::KeyCode::Tab => KeyCode::Tab,
                ct::KeyCode::Delete => KeyCode::Delete,
                ct::KeyCode::Home => KeyCode::Home,
                ct::KeyCode::End => KeyCode::End,
                ct::KeyCode::PageUp => KeyCode::PageUp,
                ct::KeyCode::PageDown => KeyCode::PageDown,
                ct::KeyCode::Esc => KeyCode::Esc,
                _ => KeyCode::Null,
            };

            let kind = match key.kind {
                ct::KeyEventKind::Press => KeyEventKind::Press,
                ct::KeyEventKind::Release => KeyEventKind::Release,
                ct::KeyEventKind::Repeat => KeyEventKind::Repeat,
            };

            KeyEvent {
                code,
                modifiers,
                kind,
            }
        }
    }
}

#[cfg(target_arch = "wasm32")]
mod wasm {
    use super::*;
    use ratzilla::event as rz;

    impl From<rz::KeyEvent> for KeyEvent {
        fn from(event: rz::KeyEvent) -> Self {
            let code = match event.code {
                rz::KeyCode::Char(c) => KeyCode::Char(c),
                rz::KeyCode::F(n) => KeyCode::F(n),
                rz::KeyCode::Backspace => KeyCode::Backspace,
                rz::KeyCode::Enter => KeyCode::Enter,
                rz::KeyCode::Left => KeyCode::Left,
                rz::KeyCode::Right => KeyCode::Right,
                rz::KeyCode::Up => KeyCode::Up,
                rz::KeyCode::Down => KeyCode::Down,
                rz::KeyCode::Tab => KeyCode::Tab,
                rz::KeyCode::Delete => KeyCode::Delete,
                rz::KeyCode::Home => KeyCode::Home,
                rz::KeyCode::End => KeyCode::End,
                rz::KeyCode::PageUp => KeyCode::PageUp,
                rz::KeyCode::PageDown => KeyCode::PageDown,
                rz::KeyCode::Esc => KeyCode::Esc,
                rz::KeyCode::Unidentified => KeyCode::Null,
            };

            KeyEvent {
                code,
                modifiers: KeyModifiers::new(event.ctrl, event.alt, event.shift),
                kind: KeyEventKind::Press,
            }
        }
    }
}

pub fn to_textarea_input(input: &Input) -> Option<TextAreaInput> {
    match input {
        Input::Key(key) => Some(key_to_textarea_input(key)),
        _ => None,
    }
}

pub fn key_to_textarea_input(key: &KeyEvent) -> TextAreaInput {
    use tui_textarea::Key;

    let key_code = match key.code {
        KeyCode::Char(c) => Key::Char(c),
        KeyCode::F(n) => Key::F(n),
        KeyCode::Backspace => Key::Backspace,
        KeyCode::Enter => Key::Enter,
        KeyCode::Left => Key::Left,
        KeyCode::Right => Key::Right,
        KeyCode::Up => Key::Up,
        KeyCode::Down => Key::Down,
        KeyCode::Tab => Key::Tab,
        KeyCode::Delete => Key::Delete,
        KeyCode::Home => Key::Home,
        KeyCode::End => Key::End,
        KeyCode::PageUp => Key::PageUp,
        KeyCode::PageDown => Key::PageDown,
        KeyCode::Esc => Key::Esc,
        KeyCode::Null => Key::Null,
    };

    TextAreaInput {
        key: key_code,
        ctrl: key.modifiers.ctrl,
        alt: key.modifiers.alt,
        shift: key.modifiers.shift,
    }
}
