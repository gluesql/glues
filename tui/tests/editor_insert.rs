#[macro_use]
mod tester;
use tester::Tester;

use color_eyre::Result;
use glues_tui::input::KeyCode;

#[tokio::test]
async fn editor_keymap_overlay_in_insert_mode() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;

    // open first note and enter insert mode
    t.open_first_note().await?;
    t.press('i').await;

    // show editor keymap (Ctrl+h)
    t.ctrl('h').await;
    t.draw()?;
    snap!(t, "editor_keymap_open");

    // any key closes overlay
    t.press('x').await;
    t.draw()?;
    snap!(t, "editor_keymap_closed");

    Ok(())
}

#[tokio::test]
async fn insert_typing_and_escape() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;
    t.open_first_note().await?;

    t.press('i').await;
    t.draw()?;
    snap!(t, "insert_mode");

    t.type_str("Hello").await;
    t.draw()?;
    snap!(t, "inserted_text");

    // escape back to normal
    t.key(KeyCode::Esc).await;
    t.draw()?;
    snap!(t, "back_to_normal");

    Ok(())
}
