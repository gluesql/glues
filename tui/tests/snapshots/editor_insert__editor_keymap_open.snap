---
source: tui/tests/editor_insert.rs
expression: text
snapshot_kind: text
---
 Note 'Sample Note' insert mode                                                                   [Ctrl+h] Show keymap 
[Browser] ┌──────────────────────────────────────────Editor Keymap───────────────────────────────────────────┐          
 󰝰 Notes  │                                                                                                  │          
   󱇗 Sampl│                                                                                                  │          
          │  | Mappings                                      | Description                                |  │          
          │  |-----------------------------------------------|--------------------------------------------|  │          
          │  | `Backspace`                                   | Delete one character before cursor         |  │          
          │  | `Ctrl+D`, `Delete`                            | Delete one character next to cursor        |  │          
          │  | `Ctrl+M`, `Enter`                             | Insert newline                             |  │          
          │  | `Ctrl+K`                                      | Delete from cursor until the end of line   |  │          
          │  | `Ctrl+J`                                      | Delete from cursor until the head of line  |  │          
          │  | `Ctrl+W`, `Alt+H`, `Alt+Backspace`            | Delete one word before cursor              |  │          
          │  | `Alt+D`, `Alt+Delete`                         | Delete one word next to cursor             |  │          
          │  | `Ctrl+U`                                      | Undo                                       |  │          
          │  | `Ctrl+R`                                      | Redo                                       |  │          
          │  | `Ctrl+C`, `Copy`                              | Copy selected text                         |  │          
          │  | `Ctrl+X`, `Cut`                               | Cut selected text                          |  │          
          │  | `Ctrl+Y`, `Paste`                             | Paste yanked text                          |  │          
          │  | `Ctrl+F`, `→`                                 | Move cursor forward by one character       |  │          
          │  | `Ctrl+B`, `←`                                 | Move cursor backward by one character      |  │          
          │  | `Ctrl+P`, `↑`                                 | Move cursor up by one line                 |  │          
          │  | `Ctrl+N`, `↓`                                 | Move cursor down by one line               |  │          
          │  | `Alt+F`, `Ctrl+→`                             | Move cursor forward by word                |  │          
          │  | `Atl+B`, `Ctrl+←`                             | Move cursor backward by word               |  │          
          │  | `Alt+]`, `Alt+P`, `Ctrl+↑`                    | Move cursor up by paragraph                |  │          
          │  | `Alt+[`, `Alt+N`, `Ctrl+↓`                    | Move cursor down by paragraph              |  │          
          │  | `Ctrl+E`, `End`, `Ctrl+Alt+F`, `Ctrl+Alt+→`   | Move cursor to the end of line             |  │          
          │  | `Ctrl+A`, `Home`, `Ctrl+Alt+B`, `Ctrl+Alt+←`  | Move cursor to the head of line            |  │          
          │  | `Alt+<`, `Ctrl+Alt+P`, `Ctrl+Alt+↑`           | Move cursor to top of lines                |  │          
          │  | `Alt+>`, `Ctrl+Alt+N`, `Ctrl+Alt+↓`           | Move cursor to bottom of lines             |  │          
          │  | `Ctrl+V`, `PageDown`                          | Scroll down by page                        |  │          
          │  | `Alt+V`, `PageUp`                             | Scroll up by page                          |  │          
          │                                                                                                  │          
          │  Thanks to tui-textarea                                                                          │          
          │                                                                                                  │          
          │                                                                                                  │          
          │                                      Press any key to close                                      │          
          │                                                                                                  │          
          └──────────────────────────────────────────────────────────────────────────────────────────────────┘          
                                            ▐ INSERT   󰝰 Notes  󱇗 Sample Note 
